### PR TITLE
Add built-in procedure for changing password

### PR DESCRIPTION
**Pull request by henriknyman**
This PR changes how Bolt handles when the authenticated user is required to change password. 
Instead of closing the session, it is kept alive to allow calling a new built-in procedure sys.auth.changePassword. Any transactions opened will have access mode NONE, which only allows access to a new type of dbms operations (currently the only operation is procedureCallDbms).

We now tie the authenticated subject that implements the AccessMode interface to a transaction, instead of having just a static AccessMode.

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/BuiltInProcedures.java [hunk 1]
```diff
@@ -52,5 +52,8 @@ public void accept( Procedures procs ) throws ProcedureException
         // functionality - eg. things that apply across databases.
         procs.register( new ListProceduresProcedure( procedureName( "sys", "procedures" ) ) );
         procs.register( new ListComponentsProcedure( procedureName( "sys", "components" ), neo4jVersion ) );
+
+        // These are 'sys.auth'-namespaced procedures; these deal with authentication and authorization-oriented operations
+        procs.register( new AlterUserPasswordProcedure( procedureName( "sys", "auth", "changePassword" ) ) );
```
Comment (boggle) [inline community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/BuiltInProcedures.java]: We talked this over in the drivers and we'd like to not introduce nested namespaces right now, please change to `sys.changePassword`

We can pick this discussion up again once we add more `auth` related functions.

File: community/kernel/src/test/java/org/neo4j/kernel/builtinprocs/BuiltInProceduresTest.java [hunk 2]
```diff
@@ -143,6 +145,7 @@ public void shouldListCorrectBuiltinProcedures() throws Throwable
             record( "db.labels", "db.labels() :: (label :: STRING?)" ),
             record( "db.propertyKeys", "db.propertyKeys() :: (propertyKey :: STRING?)" ),
             record( "db.relationshipTypes", "db.relationshipTypes() :: (relationshipType :: STRING?)" ),
+            record( "sys.auth.changePassword", "sys.auth.changePassword(password :: STRING?) :: ()" ),
```
Comment (technige) [inline community/kernel/src/test/java/org/neo4j/kernel/builtinprocs/BuiltInProceduresTest.java]: Too early to sub-namespace these. Let's stick with `sys.changePassword` for now and add more levels when we have significantly more built-ins. The presence of the word "password" already suggests this is auth-related.

File: community/common/src/main/java/org/neo4j/kernel/api/security/AccessMode.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+ */
+package org.neo4j.kernel.api.security;
+
+/** Controls the capabilities of a KernelTransaction. */
+public interface AccessMode
+{
+    enum Static implements AccessMode
+    {
+        /** No reading or writing allowed. */
+        NONE
```
Comment (technige) [inline community/common/src/main/java/org/neo4j/kernel/api/security/AccessMode.java:28]: Strictly speaking, is this `NONE` or `PASSWORD_CHANGE_ONLY`? If the latter, we should name it as such. If the former, what is the value of being allowed in with absolutely no access, as opposed to not being allowed in?

Comment (henriknyman) [inline community/common/src/main/java/org/neo4j/kernel/api/security/AccessMode.java:28]: A user being allowed to change his own password is implemented as a more fine-grained permission on the AuthSubject (which is now something like the "security context" tied to the transaction, even though the kernel statement operations currently only views it through the AccessMode interface)

File: community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/BuiltInProcedures.java (hunk 1 above)
Comment (henriknyman) [inline community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/BuiltInProcedures.java]: Done

File: community/kernel/src/test/java/org/neo4j/kernel/builtinprocs/BuiltInProceduresTest.java (hunk 2 above)
Comment (henriknyman) [inline community/kernel/src/test/java/org/neo4j/kernel/builtinprocs/BuiltInProceduresTest.java]: Done

File: community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/AlterUserPasswordProcedure.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        AccessMode accessMode = ctx.get( Context.KERNEL_TRANSACTION ).mode();
+        if ( !(accessMode instanceof AuthSubject) )
+        {
+            throw new AuthorizationViolationException( "Invalid attempt to change the password" );
+        }
+        AuthSubject authSubject = (AuthSubject) accessMode;
+
+        try
+        {
+            authSubject.setPassword( input[0].toString() );
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/AlterUserPasswordProcedure.java:66]: should we add some restrictions on what password should be?

Comment (henriknyman) [inline community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/AlterUserPasswordProcedure.java:66]: I think this should be up to the AuthManager implementation (and to throw an appropriate exception). Unfortunately the current legacy BasicAuthManager have no restrictions, and the few restrictions we have are instead implemented directly in the REST UserService.
For enterprise we will probably have requirements for password policies so I think we can deal with it later in the new auth manager.

**Pull request merged**